src/nupic/ntypes/Sdr.hpp

```diff
@@ -0,0 +1,273 @@
+ *    The outter list contains an entry for each dimension in the SDR.  The
+ *    inner lists contain the coordinates of each true bit.  The inner lists run
+ *    in parallel. This format is useful because it contains the location of
+ *    each true bit inside of the SDR's dimensional space.
+ *
+ *    Flat Index Format: A list of the indices of the true bits, indexed into
+ *    the the flattened SDR.  This format allows for quickly accessing all of
+ *    the true bits in the SDR.
+ *
+ * Flat Formats: This class uses C-order throughout, meaning that when iterating
+ * through the SDR, the last/right-most index changes fastest.
+ *
+ * Boolean Data Type: Although the documentation for this class may refer to the
+ * values in an SDR as bits they are in fact stored as bytes (of type signed
+ * character).
+ *
+ * Example usage:
+ *
+ *    // Make an SDR with 10,000 bits, arranged in a (100 x 100) gird.
+ *    SparseDistributedRepresentation X({100, 100});
+ *
+ *    // Assign data in any format.
+ *    X.setFlatIndex( mySparseData ); // Assigns mySparseData to SDR's value.
+ *    X.setDense( myDenseData );      // myDenseData overwrites mySparseData.
+ *
+ *    // Access data in different formats.
+ *    indices      = X.getFlatIndex(); // Calculated from myDenseData.
+ *    indicesAgain = X.getFlatIndex(); // Reuses previous result, no work done.
+ *
+ */
+class SparseDistributedRepresentation : public Serializable
+{
+public:
+    /**
+     * Create an uninitialized SDR.  The only way to initialize SDRs created by
+     * this method is by calling SDR.load().  Until SDR.load() is called this
+     * SDR is in an unusable state.
+     */
+    SparseDistributedRepresentation();
+
+    /**
+     * Create an SDR object.  Initially this SDR has no value set.
+     *
+     * @param dimensions A list of dimension sizes, defining the shape of the
+     * SDR.
+     */
+    SparseDistributedRepresentation(const vector<UInt> dimensions);
+
+    /**
+     * Initialize this SDR as a copy of the given SDR.  This is a deep copy
+     * meaning that all of the data in the given SDR is fully copied, and
+     * modifying either SDR will not affect the other SDR.
+     *
+     * @param value An SDR to replicate.
+     */
+    SparseDistributedRepresentation(const SparseDistributedRepresentation value);
+
+    /**
+     * @returns A list of dimensions of the SDR.
+     */
+    const vector<UInt> getDimensions() const
+        {return dimensions;};
+
+    /**
+     * @returns The total number of boolean values in the SDR.
+     */
+    const UInt getSize() const
+        {return size;};
+
+    /**
+     * Remove the value from this SDR.  Attempting to get the value of an empty
+     * SDR will raise an exception.
+     */
+    void clear();
+
+    /**
+     * Set all of the values in the SDR to false.  This method overwrites the
+     * SDRs current value.
+     */
+    void zero();
```

Comment:
We would not only like to initialize an array with 0's but also a random distribution of a given sparsity.  I don't know...maybe that is going too far.

Reply:
Good idea!  A method to create a random SDR of a given sparsity should not be too hard to make.

Another method could add random noise into an existing SDR by randomly swapping a fraction of the active & inactive bits, which yields a new SDR with a controlled amount of overlap between it and the original SDR.

Reply:
`add random noise `
Cool